### PR TITLE
osu! Difficulty Aim Overhaul

### PR DESCRIPTION
**Pull request by Xexxar**
# osu!Standard Aim Enhancement

Hello all,

This is a overhaul of the aim skill to try to address some core issues with how aim is weighted for high bpm jumps and sliders. I wanted to keep things simple, so none of the core components of geometric decay based strain systems were changed.

## Changes

The goals of this change are:

1: Clean up jank old code I wrote and set up scales and vectors that can be adjusted for future balancing.
2: Alter angle algorithm to be sensitive to deltaTime changes, adding angle insensitive buffs for patterns where deltaTime changes.
3: Reward aim with wide angles for deltaTime > 100 (slower than 300 bpm 1/2).
4: Reward aim with acute angles for deltaTime < 100 (faster than 300 bpm 1/2).
5: Scale sliders so that they actually contribute to difficulty, especially for fast or long sliders.
6: Slightly nerf spaced streams by reducing distance scaling (this is compensated by cutstream buff).
7: Buff changes in velocity when rhythm is unchanging (stream jumps)

### 1: Clean up jank

One of the ways I chose to clean up the jank with aim was to use vectors. Originally my goal was to use vector addition and subtraction for better pattern buffing, but this became unneeded. This code could be removed, but I think long term vectors are going to be better for a lot of reasons and as such I left them in. Whether they are used or not, they serve as a cleaner way to read the code.

### 2: Alter the angle algorithm

The main change here is to restrict angleBuffs to only be applied onto patterns where notes are the same deltaTime. This way, we buff angles based on what we expect to be harder for each deltaTime range, which is detailed in the next sections. For patterns where deltaTime changes, we buff those patterns automatically based on the lowest velocity compared between the two notes. This does have an additional restriction to not buff during a fast -> slow -> fast transition, as the next note will be a slow -> fast -> fast transition which will be buffed. This prevents double buffing and overscaling of these patterns. (think a stream *jump* stream).

### 3: Reward wide angles

This is basically unchanged compared to live. The only difference is that this wide angle buff begins to dramatically decrease after reaching 100 MS deltaTime (or 300 bpm 1/2). This is to prevent streams from being overbuffed. I also reduced the scaling for angles > 120 to nerf purely linear aim as this is typically easier than having some angle less than 180. A caveat of this change is that patterns where distance is uneven are typically nerfed as well (there is also a square nerf as part of this change, which is intended, though I'm not entirely sure how it happened as I didn't change the angle range for this). 

https://www.desmos.com/calculator/ubggh4ykvw 

Angle Graph for wide angles. x is angle in radians, y is multiplier.

### 4: Reward acute angles

Many maps contain very high bpm acute angles that are difficult to hit due to their rapid speed change. Currently, there is no bonus for these types of maps which is a shame, as they are typically very underweighted. This change hopes to fix this. Because we don't want to buff acute angles for lower bpm, this buff starts at 300 bpm 1/2 and scales up to approximately 400 bpm 1/2 (100->75 deltaTime). This buff scales quadratically within that range. Normally angle buffs use the lesser of the min velocity as their base, but for this we cap the velocity buff with the following code.

```cs
                            angleBonus = Math.Min(angleBonus, 150 / osuCurrObj.StrainTime)
          * Math.Min(1, Math.Pow(Math.Min(osuCurrObj.JumpDistance, osuPrevObj.JumpDistance) / 150, 2));;
```

This code scales the angleBuff to be maxed out at the distance of 150 (which is 1.5 circles worth of distance), and also nerfs distances less than this in order to prevent buffing wiggle streams that overlap. A graph of this buffs range with angles can be seen below.

https://www.desmos.com/calculator/2tx9iuywsv

Angle graph for acute angles. x is the angle in radians, y is the multiplier.

### 5: Scale sliders

This is a very simple change. In order to make sliders contribute to difficulty, I have separated them out into their own calculation and have added a buff to their distance that has a scaling factor. I also made two unique scaling factors, one for overall sliders, and one for a geometric average slider distance scale. This isn't perfect, but is a nice step forward.

### 6: Slightly Nerf Streams

This isn't really much of a nerf, as the effects are very small. In general I'm just decreasing the scaling with distance to make reaching peak tap strain harder. Some stream maps will still gain because of the acute angle buff (such as the original Sidetracked Day mapset). Technically we can nerf these further with larger spacing increases, but the small change I believe is suitable in improving the balancing.

### 7: Buff velocity changes

A bit of an afterthought, I added this in to buff maps like Polygon (Bonzi's Ultra) and Identity Part 4. These are maps where a pattern goes from being a fast jump to a slow jump or are stream jumps. This typically requires great aim control and as such are subject to buffs as they are typically underweighted due to the low velocity required to hit the slower note (this basically adds context to the motion from the previous note influencing the current).

## Conclusion

There's nothing revolutionary in this update. Just some nice QOL changes to improve balancing across the board. This combined with my rhythm changes I believe will produce a massive meta change not seen since the 2018 speed and aim rebalance. I think this is a great step forward to introducing a wider variety of maps that are valid ways of farming pp beyond the current status quo.

As always, thanks for reading.

Xexxar (he who must decrease)

### CODE REVIEW

**Comment by stanriders** [comment]
Can we have like every change listed in the post a separate PR to make sure nether of them breaks the system?

**Comment by bdach** [comment]
Code quality improvements (so (1) and maybe (2)) should definitely be on their own and be crossreferenced against current values to ensure the refactoring is in fact a refactoring and doesn't touch/change anything else. As for the rest I'm not super sure whether it needs to be split, as to me it's mostly Numbers Changing™ so I'll defer to the committee's opinion.

**Comment by Xexxar** [comment]
There really should be no need to split this up. All pp changes are coordinated in tandem and the code cleanliness / optimization aspect is simply adding vectors data to osuDifficultyHitObject and improving the code through creatung the better values. Its all addressing the same chunk of code as well.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:117]: `Math.PI / 3 - (angle - 2 * Math.PI / 3))` is a pretty convoluted way of writing `Math.PI - angle`. was this the intention here?

**Review by bdach — DISMISSED** [review]
mostly code quality stuff, the number shuffling is indecipherable to me

File: osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:126]: Remove the _slightly_ 👍

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:112]: I didnt end up using this vector piece at all if we want to remove it thats fine.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:115]: Remove this.

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs:37]: Renove this

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:31]: Yes. I put it there in case people want to adjust it on their own to try balancing for themselves. 1.0 worked for me but it may req buffs or nerfs in future.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:117]: Yeah this is pretty dumb. I was visualizing the angle change weirdly. As long as the graph is the same the change is fine.

Comment (bdach) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:117]: Actually now that you talk about it, on comparing the desmos doesn't even have the `2pi / 3` factor in. So I'm just going to remove that to match.

**Comment by bdach** [comment]
Have resolved the review above on my own at request of @Xexxar.

@Xexxar please double-check 506a58b at your leisure - it should be equivalent to the previous version when it comes to the results, but the formulas should now be easier to understand as to what they're doing.

I leave evaluating the gameplay merits of this change to the committee.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (MBmasher) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:46]: These could be renamed to `currVelocity` and `prevVelocity` for more clarity

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (MBmasher) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:84]: I think `velChangeBonus` needs some sort of angle bonus implemented, I would say velocity changes are a lot harder on wide angles.

**Comment by stanriders** [comment]
@Xexxar some things here such as velocity change buff are completely separate and were added just because they're also "aim". This can easily be split into 3 prs: refactoring/cleanup, angle buffs and other balancing, velocity change buff. Please do so so that we can be sure that neither of those parts breaks so that later on when someone decides to work with aim it wouldnt have some hidden problems that nobody knows about and that werent noticed just because it was balanced to not break.

**Comment by Xexxar** [comment]
It is not my intention to section off a holistic change to the aim assessment algorithm into seperate parts that will each increase overhead and merge complexity when the change is intended to work in total.

If there are issues with the current build that require a closer examination into the intracacies of change, then it may make sense to split this PR. But as there have been no noticably major complaints, I dont see what the advantage of splitting this PR would be. If it is purely something that you desire Stan, I encourage you to fork my branch and create the seperate PRs and I will gladly delete this one in favor of seperate proposals as you have suggested.

**Comment by stanriders** [comment]
I wouldn't call it holistic considering you've just added things on top of things. 

> will each increase overhead and merge complexity

I hope you are aware that reviewing/testing small parts separately **reduces** merge complexity. Currently people reviewing this have to basically go through this pr by small parts on their own and either assume or actually split and test changes locally to see that everything is okay and works as intended. And I'm not even going to comment on "fork and create separate PRs yourself", come on. 
I guess if mine and bdach's opinions on the subject aren't enough for you we can ask @smoogipoo or @peppy. I can, if everyone decide so, review this as a whole, but I would really prefer not to, and I would really prefer this way of presenting changes not to be a standard.

**Comment by emu1337** [comment]
Split PRs would probably be easier to check, but I'm fine either way. I'll hold off reviewing until theres an agreement.

**Comment by Xexxar** [comment]
As far as I understand it, it is not the job of the PP committee to examine the nitty gritty details of how a change is implemented or whether the code quality is sufficient, but rather it is to validate that the change at hand is a net positive improvement to the game and the ranking system at large. In a skill rework, all changes interact with eachother, as all maps have unique features that are in balance with eachother. One particular aspect of the rework when interacting with another may have unforeseen effects if not tested in aggregate. This is why I am extremely against the notion of segmenting changes, especially when they are primarily affecting a single skill like Aim. Something as simple as changing the value of a constant multiplier on a certain term we're buffing can look great in isolation until you introduce another change such that those two terms interact negatively together. Afterall, at the end of the day, all values work together to produce a final SR/PP value. 

If this change is positive, give it an approval. If its not, there should be a results based discussion on why it isn't adaquate. Its does not need to be needlessly obsfuncated by inspecting changes in isolation when the net effect of all the changed together is what we care about.

**Comment by stanriders** [comment]
If all values are so tangled up that splitting unrelated parts breaks everything then you did something wrong (which i think you didnt, because again you just added things on top of things with little balancing required). 

Role of pp committee is indeed to validate if the change is net positive for the game and I'm not sure why you assume that it can be done without examining every little detail of the proposal. Inspecting changes in isolation takes less time, is easier to confirm that the change is doing what it needs to do and by the end of it we'll get to the same combined balanced values, but knowing that these values won't explode suddenly explode because of an oversight somewhere in the beginning.

I'm done discussing this, gonna just wait for @peppy or @smoogipoo. Until then, I'll refrain from reviewing this.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:82]: applied this instead

```cs
                    double velChangeBonus = Math.Max(0, Math.Sqrt((prevVector.Length - currVector.Length) * currVector.Length) - Math.Max(0, currVector.Length - 100 / osuCurrObj.StrainTime)) * Math.Min(1, osuCurrObj.JumpDistance / 100);
```

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:46]: will let this be handled by further reviews under https://github.com/ppy/osu/pull/14845

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:84]: added this by adding (1 + wideAngleBonus) multiplier.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:82]: nvm above sucks, applied what you had.

**Comment by stanriders** [comment]
Split into #14845 #14846 #14847

**Pull request closed without merging**
